docs/decisions/0022-planner-telemetry-enhancement.md

```diff
@@ -0,0 +1,225 @@
+}
+```
+
+```csharp
+// In Connectors.AI.OpenAI ClientBase.cs
+private static readonly ActivitySource s_activitySource = new(typeof(ClientBase).FullName);
+
+// In internal GetTextResults and GetChatResults calls
+using var activity = s_activitySource.StartActivity($"{this.GetType().Name}.GetTextResults");
+if (activity?.GetBaggageItem("Operation") is not null)
+{
+    activity?.AddTag("Operation", activity.GetBaggageItem("Operation"));
+}
+```
+
+In an application that uses SK and Application Insights:
+
+```csharp
+var operations = new ConcurrentDictionary<string, IOperationHolder<DependencyTelemetry>>();
+
+// For more detailed tracing we need to attach Activity entity to Application Insights operation manually.
+void activityStarted(Activity activity)
+{
+  var operation = telemetryClient.StartOperation<DependencyTelemetry>(activity);
+  operation.Telemetry.Type = activity.Kind.ToString();
+
+  operations.TryAdd(activity.SpanId.ToString(), operation);
+}
+
+// We also need to manually stop Application Insights operation when Activity entity is stopped.
+void activityStopped(Activity activity)
+{
+  if (operations.TryRemove(activity.SpanId.ToString(), out var operation))
+  {
+    activity.Tags.ToList().ForEach(tag =>
+    {
+      // Add the tags to Application Insights telemetry operations as properties.
+      operation.Telemetry.Properties.Add(tag.Key, tag.Value);
+    });
+    telemetryClient.StopOperation(operation);
+  }
+}
+
+// Subscribe to all traces in Semantic Kernel
+activityListener.ShouldListenTo =
+  activitySource => activitySource.Name.StartsWith("Microsoft.SemanticKernel", StringComparison.Ordinal);
+
+activityListener.Sample = (ref ActivityCreationOptions<ActivityContext> _) => ActivitySamplingResult.AllData;
+activityListener.SampleUsingParentId = (ref ActivityCreationOptions<string> _) => ActivitySamplingResult.AllData;
+activityListener.ActivityStarted = activityStarted;
+activityListener.ActivityStopped = activityStopped;
+
+ActivitySource.AddActivityListener(activityListener);
+```
+
+The above produces dependency information similar to the screenshot below:
+
+<img src="./images/0022-dependency-in-application-insights-create-plan.png" alt="Example create plan dependency generated in Application Insights" width="400"/>
+
+<img src="./images/0022-dependency-in-application-insights-client-base.png" alt="Example model calling dependency generated in Application Insights" width="400"/>
+
+Cons:
+
+1. Easy to add custom data.
+2. Semi-structured data organized by tags (later saved as custom dimensions for Application Insights).
+3. Automatically correlated from parent to chid items. Baggage makes it even easier for querying.
+
+Pros:
+
+1. Everything has to a string.
+2. Developers may have to creation a new container for each activity depending on the service used.
+
+#### Log
+
+```csharp
+this._logger.LogTrace("{PlannerType}: Created plan with details: \n {Plan}", PlannerType, plan.ToPlanString());
```

Comment:
may consider using the LoggerMessage attribute which can generate optimized code. I know this is implementation details here :-)

Reply:
This PR partially addresses this comment for `KenrnelFunction` and `PlannerInstrumentation`: https://github.com/microsoft/semantic-kernel/pull/3859. Other part of the kernel will follow soon.